CMakeLists.txt

```diff
@@ -113,128 +87,70 @@ if (NOT DEFINED VERONA_LLVM_LOCATION)
 endif ()
-
-# Copied from LLVM, handling release builds with assertions isn't trivial as
-# CMake defines -DNDEBUG automatically and we have to clear it.
-if( ENABLE_ASSERTS )
-  # MSVC doesn't like _DEBUG on release builds.
-  if( NOT MSVC )
-    message(STATUS "Enabling asserts")
-    add_definitions( -D_DEBUG )
-  endif()
-  # On non-Debug builds cmake automatically defines NDEBUG, so we
-  # explicitly undefine it:
-  if( NOT uppercase_CMAKE_BUILD_TYPE STREQUAL "DEBUG" )
-    message(STATUS "Removing NDEBUG from existing flags to enable asserts")
-    # NOTE: use `add_compile_options` rather than `add_definitions` since
-    # `add_definitions` does not support generator expressions.
-    add_compile_options($<$<OR:$<COMPILE_LANGUAGE:C>,$<COMPILE_LANGUAGE:CXX>>:-UNDEBUG>)
-
-    # Also remove /D NDEBUG to avoid MSVC warnings about conflicting defines.
-    foreach (flags_var_to_scrub
-        CMAKE_CXX_FLAGS_RELEASE
-        CMAKE_CXX_FLAGS_RELWITHDEBINFO
-        CMAKE_CXX_FLAGS_MINSIZEREL
-        CMAKE_C_FLAGS_RELEASE
-        CMAKE_C_FLAGS_RELWITHDEBINFO
-        CMAKE_C_FLAGS_MINSIZEREL)
-      string (REGEX REPLACE "(^| )[/-]D *NDEBUG($| )" " "
-        "${flags_var_to_scrub}" "${${flags_var_to_scrub}}")
-    endforeach()
+set (VERONA_EXTRA_CMAKE_ARGS)
+
+# Point Verona at the MLIR
+list (APPEND VERONA_EXTRA_CMAKE_ARGS
+  -DVERONA_LLVM_LOCATION=${MLIR_INSTALL}/install
+)
+
+# Define project level defaults
+list (APPEND VERONA_EXTRA_CMAKE_ARGS
+  -DCMAKE_CXX_STANDARD=17)
+
+# Use top-level install directory required for tests.
+list (APPEND VERONA_EXTRA_CMAKE_ARGS
+  -DCMAKE_INSTALL_PREFIX=${VERONA_LOCAL_DIST})
+list (APPEND VERONA_DONT_PASS CMAKE_INSTALL_PREFIX)
+
+# Set up variables to pass arguments to sub project
+get_cmake_property(CACHE_VARS CACHE_VARIABLES)
+foreach (CACHE_VAR ${CACHE_VARS})
+  list (FIND VERONA_DONT_PASS ${CACHE_VAR} INDEX)
+  if (${INDEX} EQUAL -1)
+    list(APPEND VERONA_EXTRA_CMAKE_ARGS
+      -D${CACHE_VAR}=${${CACHE_VAR}}
+    )
   endif()
+endforeach ()
+
+ExternalProject_Add(verona
+  SOURCE_DIR ${CMAKE_SOURCE_DIR}/src
+  DEPENDS external
+  BUILD_ALWAYS true
+  USES_TERMINAL_BUILD true
+  USES_TERMINAL_CONFIGURE true
+  USES_TERMINAL_INSTALL true
+  USES_TERMINAL_TEST true
+  CMAKE_ARGS ${VERONA_EXTRA_CMAKE_ARGS})
+
+######################################################
+#  Add testing at top level
+######################################################
+enable_testing()
+
+include(ProcessorCount)
+ProcessorCount(N)
+
+add_subdirectory(testsuite)
+
+# Adds a target check that runs the tests.
+add_custom_target(check DEPENDS verona)
+add_custom_command(TARGET check 
+  COMMAND ${CMAKE_CTEST_COMMAND} -C $<CONFIG> -j ${N} --output-on-failure --timeout 400 --interactive-debug-mode 0
+  USES_TERMINAL
+)
+
+# Main target does not test runtime, build subdirectory for that
+# and add to check target.
+if (RT_TESTS)
+  add_subdirectory(src/rt)
```

Comment:
I know this was already the case before, and it probably works fine, but I find it a bit surprising that RT tests are built by the outer cmake when everything else is built by the inner one.

Reply:
The aim is to have the tests not to be part of an external project.  There are a few reasons for this.  

  1. ExternalProject with tests on MSbuild doesn't work quite right,  I was not able to build a subproject without also running its tests. This hides that fact.  
  2. It also means we have a top level ctest target, and can call it directly on particular failing tests.  Rather than having to go diving through the layers of external project to call an individual test.